### ISSUE TITLE
DateTime2 db column type should be used instead of DateTime

### ISSUE THREAD

**Issue by bantolov** [Rhetos/Rhetos#101]
**Is your feature request related to a problem? Please describe.**
For DateTime properties in an entity, Rhetos generates a `DateTime` column in the database.
DateTime column type is a legacy feature in SQL Server, DateTime2 should be used instead.
  > Use the time, date, datetime2 and datetimeoffset data types for new work. ([msdn](https://docs.microsoft.com/en-us/sql/t-sql/data-types/datetime-transact-sql?view=sql-server-2016))

We have also experienced some strange rounding errors with Entity Framework when an application tries to find a record by time (see the attached unit test [DateTimeConsistency.txt](https://github.com/Rhetos/Rhetos/files/2676692/DateTimeConsistency.txt).

**Describe the solution you'd like**
`DateTime` column type should be replaced with `DateTime2` or `DateTime2(3)`.
Currently `DateTime2(3)` seems like a better option:

OPTION 1: `DateTime2` or `DateTime2(7)` - 7 digits
* This option seems to better match the C# DateTime type.

OPTION 2: `DateTime2(3)` - milliseconds
* This option might work better for user interface implementations (JavaScript counting time in [milliseconds](https://tc39.github.io/ecma262/)). I would like to avoid a range of "roundtrip" issues when a web form loads a record and saves it without user changing some datatime field, but the change of this filed is done accidentally since the frontend has rounded the time value to milliseconds. In business application it may be an important issue noting who has and why changed a certain field value.
* Rhetos is specialized for business applications, and they rarely required precision less then millisecond. For exceptional features (high-precision measurement), some exceptional DSL concepts might be added.

**Additional notes**
1. `SYSDATETIME()` should be used instead of `GETDATE()`.
2. This feature should be implemented in a backward compatible way to avoid any issues with the existing applications. See [Backward compatible feature implementation in Rhetos](https://github.com/Rhetos/Rhetos/wiki/Backward-compatible-feature-implementation-in-Rhetos-and-CommonConcepts).
3. Add the attached unit test to *CommonConceptsTest.sln*.
4. We should write a utility SQL script for optimized migration of an existing database columns (and Rhetos metadata) from datetime to datetime2. On very large databases, this script can be used instead of a standard Rhetos upgrade process (which is robust, but slow on large data).

**Comment by bantolov** [comment]
Just another argument for `datetime2(3)` instead of `datetime2(7)`:
I have tested `SYSDATETIME()` function on my database (SQL Server 2016). In practice it does not provide higher precision then 1 ms.

```sql
CREATE TABLE #t (CreatedAt datetime2);
DECLARE @i INT = 300;
WHILE @i > 0
BEGIN
    INSERT INTO #t SELECT SYSDATETIME();
    SET @i = @i - 1;
END
SELECT CreatedAt, Count = COUNT(*) FROM #t GROUP BY CreatedAt ORDER BY 1;
DROP TABLE #t;
```

Output:

```
CreatedAt                    Count
2018-12-14 10:56:58.7777422  16
2018-12-14 10:56:58.7787400  74
2018-12-14 10:56:58.7797402  67
2018-12-14 10:56:58.7807224  56
2018-12-14 10:56:58.7817214  52
2018-12-14 10:56:58.7827257  35
```

**Comment by bantolov** [comment]
Implemented with #377 and later development. Will be released in v4.3.0.
